File: go/1-basic-programming/3-check-ab.go
package main

import(
	"fmt"
	"math"
)

func CheckAB(input string) bool {
	var (
		indexOfA []int
		indexOfB []int
		distanceThreeChar bool
	)

	for index, char := range input {
		if char == 'a' {
			indexOfA = append(indexOfA, index)
		}
		if char == 'b' {
			indexOfB = append(indexOfB, index)
		}
	}

	for _, valueIndexOfA := range indexOfA {
		for _, valueIndexOfB := range indexOfB {
			var distanceAB = math.Abs(float64(valueIndexOfA - valueIndexOfB))
			if distanceAB == 4 {
				distanceThreeChar = true
			}
		}
	}

	if distanceThreeChar {
		return true
	}
	return false
}

func main() {
	fmt.Println(CheckAB("lane borrowed")) // True
	fmt.Println(CheckAB("i am sick")) // False
	fmt.Println(CheckAB("you are boring")) // True
	fmt.Println(CheckAB("barbarian")) // True
	fmt.Println(CheckAB("bacon and meat")) // False
}

File: c++/1-basic/3-check-ab.cpp
#include <bits/stdc++.h>
using namespace std;

// complexity: O(n)
bool CheckAB(string inputString) {
  for (int i = 0; i < inputString.size() - 4; i++) {
    if (inputString[i] == 'a' && inputString[i + 4] == 'b') {
      return true;
    } else if (inputString[i] == 'b' && inputString[i + 4] == 'a') {
      return true;
    }
  }
  return false;
}

int main() {
  cout << CheckAB("lane borrowed") << endl; // True
  cout << CheckAB("i am sick") << endl; // False
  cout << CheckAB("you are boring") << endl; // True
  cout << CheckAB("barbarian") << endl; // True
  cout << CheckAB("bacon and meat") << endl; // False
}